File: src/lexer/lexer.cpp
#include "lexer.h"
#include "token/token.h"
#include "evaluator/evaluator.h"

namespace RTree
{

    Lexer::Lexer(const std::string b)
    {
        _buffer = b;
        _current = 0;
    }

    Lexer::~Lexer()
    {
    }

    char Lexer::current() const
    {
        size_t size = _buffer.length();
        if (!(size - 1))
            return -1;

        if (!(_current < size - 1))
            return -1;

        return _buffer[_current];
    }

    char Lexer::next()
    {
        size_t size = _buffer.length();
        if (!size)
            return -1;

        if (!((_current + 1) < size))
            return -1;

        return _buffer[++_current];
    }

    std::string Lexer::peek(size_t n) const
    {
        size_t size = _buffer.length();
        size_t tmp = _current;

        if ((_current + n) >= size)
            return std::string("");

        return _buffer.substr(++tmp, n);
    }

    std::vector<Token>& Lexer::scanTokens()
    {
        Evaluator eval;

        if (_buffer.empty())
            return tokens;

        // printf("Here Lexer 1\n");

        while (_current < (_buffer.length() - 1))
        {
            // printf("%d %lu\n", _current, (_buffer.length() - 1));

            eval.identifier(this, tokens);
            eval.comment(this);
            eval.punctuation(this, tokens);
            eval.literal(this, tokens);
        }

        // printf("Here Lexer 2\n");

        return tokens;
    }

}

File: src/lexer.rs
use crate::file::File;
use crate::rules::lexer;
use crate::Scanner;
use crate::Token;

pub struct Lexer {
    index: usize, // check if out of bounds
}

impl Lexer {
    pub fn new() -> Lexer {
        Lexer { index: 0 }
    }

    pub fn is_end(&self, len: usize) -> bool {
        self.index + 1 >= len
    }
}

impl Scanner<Vec<Token>, u8, File> for Lexer {
    fn current(&self, chunks: &Vec<u8>) -> Option<u8> {
        if chunks.len() <= self.index {
            return None;
        }
        Some(chunks[self.index])
    }

    fn next(&mut self, chunks: &Vec<u8>) -> Option<u8> {
        if chunks.len() > self.index {
            self.index += 1
        }

        if chunks.len() <= self.index {
            return None;
        }
        Some(chunks[self.index])
    }

    fn peek(&self, chunks: &Vec<u8>) -> Option<u8> {
        if chunks.len() <= self.index + 1 {
            return None;
        }
        Some(chunks[self.index + 1])
    }

    fn scan(&mut self, file_chunks: (&File, &Vec<u8>)) -> Vec<Token> {
        self.index = 0;
        let mut tokens: Vec<Token> = Vec::new();
        let (_file, chunks) = file_chunks;

        loop {
            match lexer::identifier(self, chunks) {
                Some(x) => tokens.push(x),
                _ => (),
            }

            match lexer::punctuation(self, chunks) {
                Some(x) => tokens.push(x),
                _ => (),
            }

            match lexer::literal(self, chunks) {
                Some(x) => tokens.push(x),
                _ => (),
            }

            lexer::comment(self, chunks);
            lexer::unknown(self, chunks);

            if self.is_end(chunks.len()) {
                break;
            }
        }

        tokens
    }
}
